jetty-core/jetty-siwe/src/main/java/org/eclipse/jetty/security/siwe/SignInWithEthereumToken.java

```diff
@@ -0,0 +1,91 @@
+// ========================================================================
+//
+
+package org.eclipse.jetty.security.siwe;
+
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.function.Predicate;
+
+import org.eclipse.jetty.security.ServerAuthException;
+import org.eclipse.jetty.util.IncludeExcludeSet;
+import org.eclipse.jetty.util.StringUtil;
+
+public record SignInWithEthereumToken(String scheme,
+                                      String domain,
+                                      String address,
+                                      String statement,
+                                      String uri,
+                                      String version,
+                                      String chainId,
+                                      String nonce,
+                                      String issuedAt,
+                                      String expirationTime,
+                                      String notBefore,
+                                      String requestId,
+                                      String resources)
+{
+
+    public void validate(SignedMessage signedMessage, Predicate<String> validateNonce,
+                         IncludeExcludeSet<String, String> schemes,
+                         IncludeExcludeSet<String, String> domains,
+                         IncludeExcludeSet<String, String> chainIds) throws ServerAuthException
+    {
+        if (validateNonce != null && !validateNonce.test(nonce()))
+            throw new ServerAuthException("invalid nonce");
+
+        if (!StringUtil.asciiEqualsIgnoreCase(signedMessage.recoverAddress(), address()))
+            throw new ServerAuthException("signature verification failed");
+
+        if (!"1".equals(version()))
+            throw new ServerAuthException("unsupported version");
+
+        LocalDateTime now = LocalDateTime.now();
+        if (StringUtil.isNotBlank(expirationTime()))
+        {
+            LocalDateTime expirationTime = LocalDateTime.parse(expirationTime(), DateTimeFormatter.ISO_DATE_TIME);
+            if (now.isAfter(expirationTime))
+                throw new ServerAuthException("expired SIWE message");
+        }
+
+        if (StringUtil.isNotBlank(notBefore()))
+        {
+            LocalDateTime notBefore = LocalDateTime.parse(notBefore(), DateTimeFormatter.ISO_DATE_TIME);
+            if (now.isBefore(notBefore))
+                throw new ServerAuthException("SIWE message not yet valid");
+        }
+
+        if (schemes != null && !schemes.test(scheme()))
+            throw new ServerAuthException("unregistered scheme");
+        if (domains != null && !domains.test(domain()))
+            throw new ServerAuthException("unregistered domain");
+        if (chainIds != null && !chainIds.test(chainId()))
+            throw new ServerAuthException("unregistered chainId");
+    }
+
+    @Override
+    public String toString()
```

Comment:
I'd leave the default implementation of `toString()`.
If you need special formatting, make an extra method such as `toMessage()`, or `asString()` with proper javadocs.

Reply:
removed, I used this to help debugging but not necessary.